contracts/provider/assembly/index.ts

```diff
@@ -0,0 +1,216 @@
+function DepositAdded(
+  depositee: string,
+  depositor: string,
+  amount: u128
+): void {
+  logging.log(
+    `DepositAdded("${depositee}", "${depositor}", "${amount.toString()}")`
+  );
+}
+
+/**
+ * Logging event fired when a deposit has been released.
+ */
+function DepositReleased(
+  depositee: string,
+  depositor: string,
+  amount: u128
+): void {
+  logging.log(
+    `DepositReleased("${depositee}", "${depositor}", "${amount.toString()}")`
+  );
+}
+
+const KEY_INITIATED = "initiated";
+
+export function initialize(
+  providerProportion: f64,
+  sessionDivisor: u128,
+  apiEndpoint: string
+): void {
+  onlyOwner();
+  if (storage.getPrimitive<bool>(KEY_INITIATED, false)) {
+    return;
+  }
+  setProviderProportion(providerProportion);
+  setSessionDivisor(sessionDivisor);
+  setApiEndpoint(apiEndpoint);
+  storage.set<bool>(KEY_INITIATED, true);
+}
+
+// Public Methods
+
+/**
+ * Deposit attached funds with this provider for the given account to initiate a session.
+ * Use `attachedDeposit` to attach a deposit in yocto-Ⓝ that will produce a session proptional
+ * in length to `sessionDivisor`.
+ */
+export function addDeposit(depositee: string): void {
+  assert(
+    context.attachedDeposit > u128.Zero,
+    "BridgeProvider: must include deposit > 0"
+  );
+  if (deposits.contains(depositee)) {
+    // Maybe it has expired?
+    _releaseDeposit(depositee);
+    assert(
+      !deposits.contains(depositee),
+      "BridgeProvider: depositee already has deposit"
+    );
+  }
+  deposits.set(
+    depositee,
+    new Deposit(context.blockTimestamp, context.sender, context.attachedDeposit)
+  );
+  DepositAdded(depositee, context.sender, context.attachedDeposit);
+}
+
+function isDepositValid(d: Deposit | null, t: u64, div: u128): bool {
+  return (
+    d != null &&
+    d.value > u128.Zero &&
+    t <= d.timestamp + u128.div(d.value, div).toI64()
+  );
+}
+
+/**
+ * Return whether the given depositee has funds deposited with this provider.
+ */
+export function hasDeposit(depositee: string): bool {
+  return isDepositValid(
+    deposits.get(depositee),
+    context.blockTimestamp,
+    _sessionDivisor()
+  );
+}
+
+function _releaseDeposit(depositee: string): void {
+  const deposit = deposits.get(depositee);
+  if (deposit == null) return;
+  if (
+    deposit.value > u128.Zero &&
+    !isDepositValid(deposit, context.blockTimestamp, _sessionDivisor())
+  ) {
+    let value = deposit.value;
+    const cut = u128.fromF64(f64.mul(value.toF64(), _providerProportion()));
+    if (cut > u128.Zero) {
+      if (value < cut) {
+        value = u128.Zero;
+      } else {
+        value = u128.sub(value, cut);
+      }
+    }
+    ContractPromiseBatch.create(deposit.depositor).transfer(value);
+    deposits.delete(depositee);
+    DepositReleased(depositee, deposit.depositor, deposit.value);
+  }
+}
+
+/**
+ * Release expired session associated with the given depositee.
+ */
+export function releaseDeposit(depositee: string): void {
```

Comment:
We wrap the more function here with a notPayable flag. And the plural version is identical to the ETH one.